File: src/main/java/com/nextfit/nutrition/service/ArticleService.java
package com.nextfit.nutrition.service;

import com.nextfit.nutrition.model.ArticleNutrition;
import com.nextfit.nutrition.model.Utilisateur;
import com.nextfit.nutrition.repository.ArticleNutritionRepository;
import com.nextfit.nutrition.repository.UtilisateurRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

/**
 * Service pour la gestion des articles et vidéos de conseils nutritionnels.
 */
@Service
@Transactional(readOnly = true)
public class ArticleService {

    private final ArticleNutritionRepository articleRepo;
    private final UtilisateurRepository utilisateurRepo;

    public ArticleService(ArticleNutritionRepository articleRepo,
                           UtilisateurRepository utilisateurRepo) {
        this.articleRepo = articleRepo;
        this.utilisateurRepo = utilisateurRepo;
    }

    /** Articles accessibles à un membre donné selon son abonnement */
    public List<ArticleNutrition> listerArticlesAccessibles(Long utilisateurId) {
        return utilisateurRepo.findById(utilisateurId)
                .filter(u -> u.getTypeAbonnement() == Utilisateur.TypeAbonnement.VIP
                          || u.getTypeAbonnement() == Utilisateur.TypeAbonnement.PREMIUM)
                .map(u -> articleRepo.findAllByOrderByDatePublicationDesc())
                .orElse(articleRepo.findByReserveVipFalseOrderByDatePublicationDesc());
    }

    /** Tous les articles publics (non-VIP) */
    public List<ArticleNutrition> listerArticlesPublics() {
        return articleRepo.findByReserveVipFalseOrderByDatePublicationDesc();
    }

    /** Filtrer par catégorie */
    public List<ArticleNutrition> listerParCategorie(ArticleNutrition.CategorieNutrition categorie) {
        return articleRepo.findByCategorieOrderByDatePublicationDesc(categorie);
    }

    /** Filtrer par type (article ou vidéo) */
    public List<ArticleNutrition> listerParType(ArticleNutrition.TypeContenu typeContenu) {
        return articleRepo.findByTypeContenuOrderByDatePublicationDesc(typeContenu);
    }

    public Optional<ArticleNutrition> trouverParId(Long id) {
        return articleRepo.findById(id);
    }

    @Transactional
    public ArticleNutrition sauvegarder(ArticleNutrition article) {
        return articleRepo.save(article);
    }

    @Transactional
    public void supprimer(Long id) {
        articleRepo.deleteById(id);
    }
}

File: Services/ArticleService.cs
using Microsoft.EntityFrameworkCore;
using NextfitNutrition.Data;
using NextfitNutrition.Models;

namespace NextfitNutrition.Services;

/// <summary>Service pour les articles et vidéos de conseils nutritionnels.</summary>
public class ArticleService
{
    private readonly NutritionDbContext _ctx;

    public ArticleService(NutritionDbContext ctx) => _ctx = ctx;

    /// <summary>Articles accessibles selon l'abonnement de l'utilisateur.</summary>
    public async Task<List<ArticleNutrition>> GetPourUtilisateurAsync(long utilisateurId)
    {
        var utilisateur = await _ctx.Utilisateurs.FindAsync(utilisateurId);
        bool estVipOuPremium = utilisateur?.Abonnement
            is TypeAbonnement.VIP or TypeAbonnement.PREMIUM;

        return estVipOuPremium
            ? await GetTousAsync()
            : await GetPublicsAsync();
    }

    public async Task<List<ArticleNutrition>> GetPublicsAsync()
        => await _ctx.Articles
            .Where(a => !a.ReserveVip)
            .OrderByDescending(a => a.DatePublication)
            .ToListAsync();

    public async Task<List<ArticleNutrition>> GetTousAsync()
        => await _ctx.Articles
            .OrderByDescending(a => a.DatePublication)
            .ToListAsync();

    public async Task<ArticleNutrition?> GetByIdAsync(long id)
        => await _ctx.Articles.FindAsync(id);

    public async Task<List<ArticleNutrition>> GetParCategorieAsync(CategorieNutrition categorie)
        => await _ctx.Articles
            .Where(a => a.Categorie == categorie)
            .OrderByDescending(a => a.DatePublication)
            .ToListAsync();

    public async Task<List<ArticleNutrition>> GetParTypeAsync(TypeContenu type)
        => await _ctx.Articles
            .Where(a => a.Format == type)
            .OrderByDescending(a => a.DatePublication)
            .ToListAsync();
}
